File: rust/src/connection/network/proto/concept.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use std::{collections::HashMap, str::FromStr};

use chrono::{DateTime, FixedOffset, NaiveDate, NaiveDateTime, TimeZone as ChronoTimeZone};
use chrono_tz::Tz;
use futures::TryFutureExt;
use itertools::Itertools;
use typedb_protocol::{
    concept,
    concept_document::{self, node::leaf::Leaf as LeafProto},
    row_entry::Entry,
    value::{datetime_tz::Timezone as TimezoneProto, Value as ValueProtoInner},
    value_type::ValueType as ValueTypeProto,
    Attribute as AttributeProto, AttributeType as AttributeTypeProto, Concept as ConceptProto,
    ConceptDocument as ConceptDocumentProto, ConceptRow as ConceptRowProto, Entity as EntityProto,
    EntityType as EntityTypeProto, Relation as RelationProto, RelationType as RelationTypeProto,
    RoleType as RoleTypeProto, Value as ValueProto, ValueType as ValueTypeStructProto,
};

use super::{FromProto, TryFromProto};
use crate::{
    answer::concept_document::{Leaf, Node},
    concept::{
        value::{Decimal, TimeZone},
        Attribute, AttributeType, Concept, Entity, EntityType, Kind, Relation, RelationType, RoleType, Value,
        ValueType,
    },
    error::{
        ConnectionError,
        ConnectionError::{
            ListsNotImplemented, MissingResponseField, ValueStructNotImplemented, ValueTimeZoneNameNotRecognised,
            ValueTimeZoneOffsetNotRecognised,
        },
    },
    Error, Result,
};

impl TryFromProto<ConceptRowProto> for Vec<Option<Concept>> {
    fn try_from_proto(proto: ConceptRowProto) -> Result<Self> {
        let mut row = Vec::new();
        for entry in proto.row {
            match entry.entry.ok_or(Error::from(MissingResponseField { field: "ConceptRow.entry" }))? {
                Entry::Empty(_) => row.push(None),
                Entry::Concept(concept_proto) => row.push(Some(Concept::try_from_proto(concept_proto)?)),
                Entry::Value(value_proto) => row.push(Some(Concept::Value(Value::try_from_proto(value_proto)?))),
                Entry::ConceptList(_) | Entry::ValueList(_) => {
                    return Err(ListsNotImplemented.into());
                }
            }
        }
        Ok(row)
    }
}

impl TryFromProto<ConceptProto> for Concept {
    fn try_from_proto(proto: ConceptProto) -> Result<Self> {
        let ConceptProto { concept: concept_proto } = proto;
        match concept_proto {
            Some(concept::Concept::EntityType(entity_type_proto)) => {
                Ok(Self::EntityType(EntityType::from_proto(entity_type_proto)))
            }
            Some(concept::Concept::RelationType(relation_type_proto)) => {
                Ok(Self::RelationType(RelationType::from_proto(relation_type_proto)))
            }
            Some(concept::Concept::AttributeType(attribute_type_proto)) => {
                Ok(Self::AttributeType(AttributeType::from_proto(attribute_type_proto)))
            }

            Some(concept::Concept::RoleType(role_type_proto)) => {
                Ok(Self::RoleType(RoleType::from_proto(role_type_proto)))
            }

            Some(concept::Concept::Entity(entity_proto)) => Entity::try_from_proto(entity_proto).map(Self::Entity),
            Some(concept::Concept::Relation(relation_proto)) => {
                Relation::try_from_proto(relation_proto).map(Self::Relation)
            }
            Some(concept::Concept::Attribute(attribute_proto)) => {
                Attribute::try_from_proto(attribute_proto).map(Self::Attribute)
            }

            // Some(concept::Concept::Value(value_proto)) => Value::try_from_proto(value_proto).map(Self::Value),
            None => Err(MissingResponseField { field: "concept" }.into()),
        }
    }
}

impl TryFromProto<concept_document::Node> for Node {
    fn try_from_proto(proto: concept_document::Node) -> Result<Self> {
        match proto.node {
            Some(concept_document::node::Node::Map(map)) => Ok(Self::Map(HashMap::try_from_proto(map)?)),
            Some(concept_document::node::Node::List(list)) => Ok(Self::List(Vec::try_from_proto(list)?)),
            Some(concept_document::node::Node::Leaf(leaf)) => {
                let result: Result<Option<Leaf>> = Option::try_from_proto(leaf);
                Ok(Self::Leaf(result?))
            }
            None => Err(MissingResponseField { field: "node" }.into()),
        }
    }
}

impl TryFromProto<ConceptDocumentProto> for Option<Node> {
    fn try_from_proto(proto: ConceptDocumentProto) -> Result<Self> {
        let ConceptDocumentProto { root: root_proto } = proto;
        match root_proto {
            None => Ok(None),
            Some(node_proto) => Node::try_from_proto(node_proto).map(Some),
        }
    }
}

impl TryFromProto<concept_document::node::Map> for HashMap<String, Node> {
    fn try_from_proto(proto: concept_document::node::Map) -> Result<Self> {
        let concept_document::node::Map { map } = proto;
        map.into_iter().map(|(var, node_proto)| Node::try_from_proto(node_proto).map(|node| (var, node))).try_collect()
    }
}

impl TryFromProto<concept_document::node::List> for Vec<Node> {
    fn try_from_proto(proto: concept_document::node::List) -> Result<Self> {
        let concept_document::node::List { list } = proto;
        list.into_iter().map(Node::try_from_proto).try_collect()
    }
}

impl TryFromProto<concept_document::node::Leaf> for Option<Leaf> {
    fn try_from_proto(proto: concept_document::node::Leaf) -> Result<Self> {
        match proto.leaf {
            Some(LeafProto::Empty(_)) => Ok(None),
            Some(LeafProto::EntityType(entity_type_proto)) => {
                Ok(Some(Leaf::Concept(Concept::EntityType(EntityType::from_proto(entity_type_proto)))))
            }
            Some(LeafProto::RelationType(relation_type_proto)) => {
                Ok(Some(Leaf::Concept(Concept::RelationType(RelationType::from_proto(relation_type_proto)))))
            }
            Some(LeafProto::AttributeType(attribute_type_proto)) => {
                Ok(Some(Leaf::Concept(Concept::AttributeType(AttributeType::from_proto(attribute_type_proto)))))
            }
            Some(LeafProto::RoleType(role_type_proto)) => {
                Ok(Some(Leaf::Concept(Concept::RoleType(RoleType::from_proto(role_type_proto)))))
            }
            Some(LeafProto::ValueType(value_type_proto)) => {
                ValueType::try_from_proto(value_type_proto).map(|value_type| Some(Leaf::ValueType(value_type)))
            }
            Some(LeafProto::Attribute(attribute_proto)) => {
                Attribute::try_from_proto(attribute_proto).map(|attr| Some(Leaf::Concept(Concept::Attribute(attr))))
            }
            Some(LeafProto::Value(value_proto)) => {
                Value::try_from_proto(value_proto).map(|value| Some(Leaf::Concept(Concept::Value(value))))
            }
            Some(LeafProto::Kind(kind_proto)) => Kind::try_from_proto(kind_proto).map(|kind| Some(Leaf::Kind(kind))),
            None => Ok(None),
        }
    }
}

impl FromProto<EntityTypeProto> for EntityType {
    fn from_proto(proto: EntityTypeProto) -> Self {
        let EntityTypeProto { label } = proto;
        Self { label }
    }
}

impl FromProto<RelationTypeProto> for RelationType {
    fn from_proto(proto: RelationTypeProto) -> Self {
        let RelationTypeProto { label } = proto;
        Self { label }
    }
}

impl FromProto<AttributeTypeProto> for AttributeType {
    fn from_proto(proto: AttributeTypeProto) -> Self {
        let AttributeTypeProto { label, value_type } = proto;
        let value_type = match value_type {
            None => None,
            Some(proto) => Some(ValueType::from_proto(proto.value_type.unwrap())),
        };
        Self { label, value_type }
    }
}

impl FromProto<RoleTypeProto> for RoleType {
    fn from_proto(proto: RoleTypeProto) -> Self {
        let RoleTypeProto { label } = proto;
        Self { label }
    }
}

impl FromProto<ValueTypeProto> for ValueType {
    fn from_proto(proto: ValueTypeProto) -> Self {
        match proto {
            ValueTypeProto::Boolean(_) => Self::Boolean,
            ValueTypeProto::Long(_) => Self::Long,
            ValueTypeProto::Double(_) => Self::Double,
            ValueTypeProto::String(_) => Self::String,
            ValueTypeProto::Decimal(_) => Self::Decimal,
            ValueTypeProto::Date(_) => Self::Date,
            ValueTypeProto::Datetime(_) => Self::Datetime,
            ValueTypeProto::DatetimeTz(_) => Self::DatetimeTZ,
            ValueTypeProto::Duration(_) => Self::Duration,
            ValueTypeProto::Struct(typedb_protocol::value_type::Struct { name }) => Self::Struct(name),
        }
    }
}

impl TryFromProto<ValueTypeStructProto> for ValueType {
    fn try_from_proto(proto: ValueTypeStructProto) -> Result<Self> {
        let ValueTypeStructProto { value_type: value_type_proto } = proto;
        Ok(ValueType::from_proto(value_type_proto.ok_or(MissingResponseField { field: "value_type" })?))
    }
}

impl TryFromProto<EntityProto> for Entity {
    fn try_from_proto(proto: EntityProto) -> Result<Self> {
        let EntityProto { iid, entity_type } = proto;
        Ok(Self { iid: iid.into(), type_: entity_type.map(EntityType::from_proto) })
    }
}

impl TryFromProto<RelationProto> for Relation {
    fn try_from_proto(proto: RelationProto) -> Result<Self> {
        let RelationProto { iid, relation_type } = proto;
        Ok(Self { iid: iid.into(), type_: relation_type.map(RelationType::from_proto) })
    }
}

impl TryFromProto<AttributeProto> for Attribute {
    fn try_from_proto(proto: AttributeProto) -> Result<Self> {
        let AttributeProto { iid, attribute_type, value } = proto;
        let type_ = attribute_type.map(AttributeType::from_proto);
        Ok(Self {
            iid: iid.into(),
            type_,
            value: Value::try_from_proto(value.ok_or(MissingResponseField { field: "value" })?)?,
        })
    }
}

impl TryFromProto<ValueProto> for Value {
    fn try_from_proto(proto: ValueProto) -> Result<Self> {
        match proto.value {
            Some(ValueProtoInner::Boolean(boolean)) => Ok(Self::Boolean(boolean)),
            Some(ValueProtoInner::Long(long)) => Ok(Self::Long(long)),
            Some(ValueProtoInner::Double(double)) => Ok(Self::Double(double)),
            Some(ValueProtoInner::String(string)) => Ok(Self::String(string)),
            Some(ValueProtoInner::Decimal(decimal)) => {
                Ok(Self::Decimal(Decimal::new(decimal.integer, decimal.fractional)))
            }
            Some(ValueProtoInner::Date(date)) => {
                Ok(Self::Date(NaiveDate::from_num_days_from_ce_opt(date.num_days_since_ce).unwrap()))
            }
            Some(ValueProtoInner::Datetime(datetime)) => {
                Ok(Self::Datetime(naive_datetime_from_timestamp(datetime.seconds, datetime.nanos)))
            }
            Some(ValueProtoInner::DatetimeTz(datetime_tz)) => {
                let datetime = datetime_tz
                    .datetime
                    .ok_or(Error::from(MissingResponseField { field: "Value.datetime_tz.datetime" }))?;
                let naive_datetime = naive_datetime_from_timestamp(datetime.seconds, datetime.nanos);
                let timezone = datetime_tz
                    .timezone
                    .ok_or(Error::from(MissingResponseField { field: "Value.datetime_tz.timezone" }))?;

                let time_zone = match timezone {
                    TimezoneProto::Named(name) => {
                        let tz = Tz::from_str(&name)
                            .map_err(|_| Error::from(ValueTimeZoneNameNotRecognised { time_zone: name.to_owned() }))?;
                        TimeZone::IANA(tz)
                    }
                    TimezoneProto::Offset(offset_seconds) => {
                        let fixed_offset = if offset_seconds >= 0 {
                            FixedOffset::east_opt(offset_seconds)
                        } else {
                            FixedOffset::west_opt(-offset_seconds)
                        }
                        .ok_or(Error::from(ValueTimeZoneOffsetNotRecognised { offset: offset_seconds }))?;
                        TimeZone::Fixed(fixed_offset)
                    }
                };
                Ok(Self::DatetimeTZ(time_zone.from_utc_datetime(&naive_datetime)))
            }
            Some(ValueProtoInner::Duration(duration)) => {
                Ok(Self::Duration(crate::concept::value::Duration::new(duration.months, duration.days, duration.nanos)))
            }
            Some(ValueProtoInner::Struct(_struct)) => {
                Err(ValueStructNotImplemented.into())
                // TODO: not implemented yet
            }
            None => Err(MissingResponseField { field: "Value.value" }.into()),
        }
    }
}

impl TryFromProto<i32> for Kind {
    fn try_from_proto(kind: i32) -> Result<Self> {
        match kind {
            0 => Ok(Self::Entity),
            1 => Ok(Self::Relation),
            2 => Ok(Self::Attribute),
            3 => Ok(Self::Role),
            _ => Err(ConnectionError::UnexpectedKind { kind }.into()),
        }
    }
}

fn naive_datetime_from_timestamp(seconds: i64, nanos: u32) -> NaiveDateTime {
    DateTime::from_timestamp(seconds, nanos).unwrap().naive_utc()
}

File: python/typedb/api/concept/concept.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from abc import ABC, abstractmethod
from datetime import date
from decimal import Decimal
from typing import TYPE_CHECKING, Dict, Optional, Union

from typedb.common.datetime import Datetime
from typedb.common.duration import Duration
from typedb.common.exception import TypeDBDriverException, INVALID_CONCEPT_CASTING

if TYPE_CHECKING:
    from typedb.api.concept.instance.attribute import Attribute
    from typedb.api.concept.instance.entity import Entity
    from typedb.api.concept.instance.relation import Relation
    from typedb.api.concept.instance.instance import Instance
    from typedb.api.concept.type.attribute_type import AttributeType
    from typedb.api.concept.type.entity_type import EntityType
    from typedb.api.concept.type.relation_type import RelationType
    from typedb.api.concept.type.role_type import RoleType
    from typedb.api.concept.type.type import Type
    from typedb.api.concept.value.value import Value


class Concept(ABC):
    STRUCT = Dict[str, Optional["Value"]]
    VALUE = Union[bool, int, float, Decimal, str, date, Datetime, Duration, STRUCT]

    def is_type(self) -> bool:
        """
        Checks if the concept is a ``Type``.

        :return:

        Examples
        --------
        ::

            concept.is_type()
        """
        return False

    def is_entity_type(self) -> bool:
        """
        Checks if the concept is an ``EntityType``.

        :return:

        Examples
        --------
        ::

            concept.is_entity_type()
        """
        return False

    def is_attribute_type(self) -> bool:
        """
        Checks if the concept is an ``AttributeType``.

        :return:

        Examples
        --------
        ::

            concept.is_attribute_type()
        """
        return False

    def is_relation_type(self) -> bool:
        """
        Checks if the concept is a ``RelationType``.

        :return:

        Examples
        --------
        ::

            concept.is_relation_type()
        """
        return False

    def is_role_type(self) -> bool:
        """
        Checks if the concept is a ``RoleType``.

        :return:

        Examples
        --------
        ::

            concept.is_role_type()
        """
        return False

    def is_instance(self) -> bool:
        """
        Checks if the concept is a ``Instance``.

        :return:

        Examples
        --------
        ::

            concept.is_instance()
        """
        return False

    def is_entity(self) -> bool:
        """
        Checks if the concept is an ``Entity``.

        :return:

        Examples
        --------
        ::

            concept.is_entity()
        """
        return False

    def is_attribute(self) -> bool:
        """
        Checks if the concept is an ``Attribute``.

        :return:

        Examples
        --------
        ::

            concept.is_attribute()
        """
        return False

    def is_relation(self) -> bool:
        """
        Checks if the concept is a ``Relation``.

        :return:

        Examples
        --------
        ::

            concept.is_relation()
        """
        return False

    def is_value(self) -> bool:
        """
        Checks if the concept is a ``Value``.

        :return:

        Examples
        --------
        ::

            concept.is_value()
        """
        return False

    def as_type(self) -> Type:
        """
        Casts the concept to ``Type``.

        :return:

        Examples
        --------
        ::

            concept.as_type()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Type"))

    def as_entity_type(self) -> EntityType:
        """
        Casts the concept to ``EntityType``.

        :return:

        Examples
        --------
        ::

            concept.as_entity_type()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "EntityType"))

    def as_attribute_type(self) -> AttributeType:
        """
        Casts the concept to ``AttributeType``.

        :return:

        Examples
        --------
        ::

            concept.as_attribute_type()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "AttributeType"))

    def as_relation_type(self) -> RelationType:
        """
        Casts the concept to ``RelationType``.

        :return:

        Examples
        --------
        ::

            concept.as_relation_type()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "RelationType"))

    def as_role_type(self) -> RoleType:
        """
        Casts the concept to ``RoleType``.

        :return:

        Examples
        --------
        ::

            concept.as_role_type()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "RoleType"))

    def as_instance(self) -> Instance:
        """
        Casts the concept to ``Instance``.

        :return:

        Examples
        --------
        ::

            concept.as_instance()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Instance"))

    def as_entity(self) -> Entity:
        """
        Casts the concept to ``Entity``.

        :return:

        Examples
        --------
        ::

            concept.as_entity()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Entity"))

    def as_attribute(self) -> Attribute:
        """
        Casts the concept to ``Attribute``.

        :return:

        Examples
        --------
        ::

            concept.as_attribute()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Attribute"))

    def as_relation(self) -> Relation:
        """
        Casts the concept to ``Relation``.

        :return:

        Examples
        --------
        ::

            concept.as_relation()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Relation"))

    def as_value(self) -> Value:
        """
        Casts the concept to ``Value``.

        :return:

        Examples
        --------
        ::

            concept.as_value()
        """
        raise TypeDBDriverException(INVALID_CONCEPT_CASTING, (self.__class__.__name__, "Value"))

    @abstractmethod
    def is_boolean(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``boolean``
        or if this ``Concept`` is an ``AttributeType`` of type ``boolean``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_boolean()
        """
        pass

    @abstractmethod
    def is_integer(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``integer``
        or if this ``Concept`` is an ``AttributeType`` of type ``integer``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_integer()
        """
        pass

    @abstractmethod
    def is_double(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``double``
        or if this ``Concept`` is an ``AttributeType`` of type ``double``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_double()
        """
        pass

    @abstractmethod
    def is_decimal(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``decimal``
        or if this ``Concept`` is an ``AttributeType`` of type ``decimal``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_decimal()
        """
        pass

    @abstractmethod
    def is_string(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``string``
        or if this ``Concept`` is an ``AttributeType`` of type ``string``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_string()
        """
        pass

    @abstractmethod
    def is_date(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``date``
        or if this ``Concept`` is an ``AttributeType`` of type ``date``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_date()
        """
        pass

    @abstractmethod
    def is_datetime(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``datetime``
        or if this ``Concept`` is an ``AttributeType`` of type ``datetime``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_datetime()
        """
        pass

    @abstractmethod
    def is_datetime_tz(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``datetime-tz``
        or if this ``Concept`` is an ``AttributeType`` of type ``datetime-tz``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_datetime_tz()
        """
        pass

    @abstractmethod
    def is_duration(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``duration``
        or if this ``Concept`` is an ``AttributeType`` of type ``duration``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_duration()
        """
        pass

    @abstractmethod
    def is_struct(self) -> bool:
        """
        Returns ``True`` if the value which this ``Concept`` holds is of type ``struct``
        or if this ``Concept`` is an ``AttributeType`` of type ``struct``.
        Otherwise, returns ``False``.

        :return:

        Examples
        --------
        ::

            concept.is_struct()
        """
        pass

    # TODO: Could be useful to have is_struct(struct_name)

    @abstractmethod
    def try_get_boolean(self) -> Optional[bool]:
        """
        Returns a ``boolean`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_boolean()
        """
        pass

    @abstractmethod
    def try_get_integer(self) -> Optional[int]:
        """
        Returns a ``integer`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_integer()
        """
        pass

    @abstractmethod
    def try_get_double(self) -> Optional[float]:
        """
        Returns a ``double`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_double()
        """
        pass

    @abstractmethod
    def try_get_decimal(self) -> Optional[Decimal]:
        """
        Returns a ``decimal`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_decimal()
        """
        pass

    @abstractmethod
    def try_get_string(self) -> Optional[str]:
        """
        Returns a ``string`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_string()
        """
        pass

    @abstractmethod
    def try_get_date(self) -> Optional[date]:
        """
        Returns a timezone naive ``date`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_date()
        """
        pass

    @abstractmethod
    def try_get_datetime(self) -> Optional[Datetime]:
        """
        Returns a timezone naive ``datetime`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_datetime()
        """
        pass

    @abstractmethod
    def try_get_datetime_tz(self) -> Optional[Datetime]:
        """
        Returns a timezone naive ``datetime_tz`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_datetime_tz()
        """
        pass

    @abstractmethod
    def try_get_duration(self) -> Optional[Duration]:
        """
        Returns a timezone naive ``duration`` value of this ``Concept``.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_duration()
        """
        pass

    @abstractmethod
    def try_get_struct(self) -> Optional[STRUCT]:
        """
        Returns a ``struct`` value of this ``Concept`` represented as a map from field names to values.
        If it's not a ``Value`` or it has another type, raises an exception.

        :return:

        Examples
        --------
        ::

            value.try_get_struct()
        """
        pass

    @abstractmethod
    def get_label(self) -> str:
        """
        Get the label of the concept.
        If this is an ``Instance``, return the label of the type of this instance ("unknown" if type fetching is disabled).
        If this is a ``Value``, return the label of the value type of the value.
        If this is a ``Type``, return the label of the type.

        :return:

        Examples
        --------
        ::

            concept.get_label()
        """
        pass

    @abstractmethod
    def try_get_label(self) -> Optional[str]:
        """
        Get the label of the concept.
        If this is an ``Instance``, return the label of the type of this instance (``None`` if type fetching is disabled).
        Returns ``None`` if type fetching is disabled.
        If this is a ``Value``, return the label of the value type of the value.
        If this is a ``Type``, return the label of the type.

        :return:

        Examples
        --------
        ::

            concept.try_get_label()
        """
        pass

    @abstractmethod
    def try_get_iid(self) -> Optional[str]:
        """
        Retrieves the unique id of the ``Concept``. Returns ``None`` if absent.

        :return:

        Examples
        --------
        ::

            concept.try_get_iid()
        """
        pass

    @abstractmethod
    def try_get_value_type(self) -> Optional[str]:
        """
        Retrieves the ``str` describing the value type fo this ``Concept``. Returns ``None`` if absent.

        :return:

        Examples
        --------
        ::

            concept.try_get_value_type()
        """
        pass

    @abstractmethod
    def try_get_value(self) -> Optional[VALUE]:
        """
        Retrieves the value which this ``Concept`` holds. Returns ``None`` if this ``Concept`` does not hold any value.

        :return:

        Examples
        --------
        ::

            concept.try_get_value()
        """
        pass
